### PR TITLE
Improve Kokkos unit testing

### PR DESCRIPTION
**Pull request by alphataubio**
**Summary**

add kokkos unit tests for angles, bonds, dihedral, improper, fixes, ... 

**Related Issue(s)**

n/a

**Author(s)**

@alphataubio 

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

n/a

**Implementation Notes**

pair_coul_dsf_kokkos INTENTIONALY SKIPPED IN YAML FILE
pair_zbl_kokkos INTENTIONALY SKIPPED IN YAML FILE

**Post Submission Checklist**

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [x] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

### CODE REVIEW

**Comment by alphataubio** [comment]
98% tests passed, 13 tests failed out of 573

The following tests FAILED:
	452 - FixTimestep:deform (Failed)
	454 - FixTimestep:deform_tri (Failed)
	456 - FixTimestep:efield_const (Failed)
	457 - FixTimestep:efield_region (Failed)
	460 - FixTimestep:efield_variable (Failed)
	472 - FixTimestep:nph (Failed)
	475 - FixTimestep:npt_iso (Failed)
	514 - FixTimestep:setforce_const (Failed)
	515 - FixTimestep:setforce_region (Failed)
	516 - FixTimestep:setforce_variable (Failed)
	524 - FixTimestep:spring_self (Failed)
	534 - FixTimestep:wall_lj93_const (Failed)
	548 - DihedralStyle:hybrid (Failed)

File: unittest/force-styles/tests/dihedral-charmmfsw.yaml
Comment (akohlmey) [inline unittest/force-styles/tests/dihedral-charmmfsw.yaml:1]: Please roll back these changes. The purpose of the test is to test the dihedral style and it does so with the minimum set of external dependencies. If you absolutely insist on testing the charmm dihedral style with a coul/long pair style, then please submit it as an *additional* test.

Comment (alphataubio) [reply unittest/force-styles/tests/dihedral-charmmfsw.yaml:1]: ok i reverted the old test and added dihedral-charmmfsw-kokkos.yaml instead. as of 2024, lammps input scripts generated by charmmgui use:

`pair_style      lj/charmmfsw/coul/long 10 12`
`dihedral_style  charmmfsw`

i added dihedral_charmmfsw_kokkos a few months ago, and i have to test it in kokkos using coul/long because pair_style lj/charmmfsw/coul/charmmfsh is not available in kokkos so the kokkos unit test for dihedral_style  charmmfsw fails.

**Comment by akohlmey** [comment]
@alphataubio This may have a merge conflict when I get to merge PR #4203 (which will be very soon).
My suggestion is that you pull that branch right away (I don't expect to make more changes to the fix style tester program after modernizing its fix pointer access and including changes to detect issues with run style respa) and then resolve the test failures. We're planning to have a feature release in a couple days and if this does not get ready and included by then, it will be postponed until after the stable release (possibly in august).

**Comment by stanmoore1** [comment]
@alphataubio @alphataubio I've reviewed the code in this PR and have a few minor change requests/comments, will submit soon. But otherwise it looks good I think it can be merged soon. @alphataubio Also is there an example I can use to test the new styles?

**Comment by stanmoore1** [comment]
Sorry, posted above comment on wrong PR, moving...

**Comment by alphataubio** [comment]
there's no rush to merge this (NOT READY). i improved the unit test coverage for KOKKOS from 34% to 43% but there are now tests failing so it should NOT be merged.

i put up this draft pr so i can see automated coverage reports in codecov without having to setup/run coverage tools on my own machine.

**Comment by stanmoore1** [comment]
> i improved the unit test coverage for KOKKOS from 34% to 43%

Nice!

**Comment by alphataubio** [comment]
> Nice!

until you find out that 13 tests are now failing ...

File: src/KOKKOS/fix_efield_kokkos.cpp [hunk 1]
```diff
@@ -43,13 +44,17 @@ FixEfieldKokkos<DeviceType>::FixEfieldKokkos(LAMMPS *lmp, int narg, char **arg)
 {
   kokkosable = 1;
   atomKK = (AtomKokkos *) atom;
+  domainKK = (DomainKokkos *) domain;
   execution_space = ExecutionSpaceFromDevice<DeviceType>::space;
-  datamask_read = EMPTY_MASK;
```
Comment (stanmoore1) [inline src/KOKKOS/fix_efield_kokkos.cpp:47]: If there is only one parallel region then it is fine to set `datamask_read` to a global value, but if there are multiple parallel regions and each need different data, then it is better to set `datamask_read = EMPTY_MASK` and then sync/modify the correct data in each function instead. This avoids transferring extra data that is not needed. So this change needs to be reverted.

File: src/KOKKOS/fix_efield_kokkos.cpp [hunk 2]
```diff
@@ -167,144 +165,186 @@ void FixEfieldKokkos<DeviceType>::post_force(int /*vflag*/)
...
+  }
+
+  if (vflag_atom) {
+    k_vatom.template modify<DeviceType>();
+    k_vatom.template sync<LMPHostType>();
   }
 
-  atomKK->modified(execution_space, F_MASK);
 
-  fsum[0] = fsum_kk.d0;
```
Comment (stanmoore1) [inline src/KOKKOS/fix_efield_kokkos.cpp:242]: @alphataubio I don't understand why this doesn't work, I think it should. The value from a parallel_reduce is automatically transferred to host CPU, it should be current.

File: src/KOKKOS/fix_nh_kokkos.cpp [hunk 3]
```diff
@@ -289,7 +295,7 @@ void FixNHKokkos<DeviceType>::remap()
   double oldlo,oldhi;
   double expfac;
 
-  int nlocal = atom->nlocal;
+  int nlocal = atomKK->nlocal;
```
Comment (stanmoore1) [inline src/KOKKOS/fix_nh_kokkos.cpp]: Not necessary, the variable is the same.

File: src/KOKKOS/fix_wall_lj93_kokkos.cpp [hunk 4]
```diff
@@ -29,11 +40,156 @@ FixWallLJ93Kokkos<DeviceType>::FixWallLJ93Kokkos(LAMMPS *lmp, int narg, char **a
   kokkosable = 1;
   atomKK = (AtomKokkos *) atom;
   execution_space = ExecutionSpaceFromDevice<DeviceType>::space;
-  datamask_read = EMPTY_MASK;
-  datamask_modify = EMPTY_MASK;
-  virial_global_flag = virial_peratom_flag = 0;
+  datamask_read = X_MASK | V_MASK | MASK_MASK;
```
Comment (stanmoore1) [inline src/KOKKOS/fix_wall_lj93_kokkos.cpp:40]: Same comment as above.

File: src/KOKKOS/fix_wall_lj93_kokkos.cpp [hunk 5]
```diff
@@ -29,11 +40,156 @@ FixWallLJ93Kokkos<DeviceType>::FixWallLJ93Kokkos(LAMMPS *lmp, int narg, char **a
...
+  memoryKK->destroy_kokkos(k_ewall,ewall);
+
+}
+
+/* ---------------------------------------------------------------------- */
+
+template <class DeviceType>
+void FixWallLJ93Kokkos<DeviceType>::precompute(int m)
+{
+  d_coeff1[m] = 6.0 / 5.0 * d_epsilon[m] * powint(d_sigma[m], 9);
```
Comment (stanmoore1) [inline src/KOKKOS/fix_wall_lj93_kokkos.cpp]: Will crash on GPU since it is host CPU code referencing GPU memory.

File: src/KOKKOS/fix_wall_lj93_kokkos.cpp [hunk 6]
```diff
@@ -42,50 +198,96 @@ FixWallLJ93Kokkos<DeviceType>::FixWallLJ93Kokkos(LAMMPS *lmp, int narg, char **a
...
+
   copymode = 1;
-  FixWallLJ93KokkosFunctor<DeviceType> wp_functor(this);
-  Kokkos::parallel_reduce(nlocal,wp_functor,ewall);
+  FixWallLJ93KokkosFunctor<DeviceType> functor(this);
+  Kokkos::parallel_reduce(nlocal,functor,result);
   copymode = 0;
 
-  atomKK->modified(execution_space, F_MASK);
+  Kokkos::atomic_add(&(d_ewall[0]),result[0]);
```
Comment (stanmoore1) [inline src/KOKKOS/fix_wall_lj93_kokkos.cpp]: Will crash on GPU.

File: src/fix_wall.cpp [hunk 7]
```diff
@@ -44,6 +42,7 @@ FixWall::FixWall(LAMMPS *lmp, int narg, char **arg) : Fix(lmp, narg, arg), nwall
   virial_global_flag = virial_peratom_flag = 1;
   respa_level_support = 1;
   ilevel_respa = 0;
+  ewall = new double[7];
```
Comment (stanmoore1) [inline src/fix_wall.cpp]: Should not be necessary.

**Review by stanmoore1 — CHANGES_REQUESTED** [review]
@alphataubio thank you for working on this. For `fix_deform_kokkos` there is no GPU device code here, so there is no reason to template on `DeviceType`, mark functions as `KOKKOS_INLINE_FUNCTION`, or use a DualView for `set`. This is because there are no loops over atoms, only much smaller loops that wouldn't run well on a GPU, so it is all host CPU code, no Kokkos `parallel_for` or `parallel_reduce`. There is also no need to change `Domain` to `DomainKK` everywhere since most of the variables inherit from the base class. I think there are just a few changes that were done to the base class that need to be propagated to the Kokkos version to get the tests to pass.

For fix efield and wall_lj93, I don't understand why the parallel_reduce with custom struct isn't working (it should be). See the comments I made in the code. Let me know if you have specific questions either here or on Slack.

File: src/KOKKOS/fix_efield_kokkos.cpp (hunk 1 above)
Comment (alphataubio) [reply src/KOKKOS/fix_efield_kokkos.cpp:47]: see if you like this commit https://github.com/lammps/lammps/pull/4190/commits/88a32e691638c93bfec22fa7d59f5d75224484bb 

By refactoring Domain and DomainKokkos with some additional utility methods and using polymorphism, fix_deform_kokkos.cpp is down to 9 lines of code and resilient to future changes in base class.

The reason for this bug was base class filtered by groupbit and kokkos class did not for x2lamda() and lamda2x() in end_of_step().

File: src/KOKKOS/fix_nh_kokkos.cpp (hunk 3 above)
Comment (alphataubio) [reply src/KOKKOS/fix_nh_kokkos.cpp]: done

File: src/KOKKOS/fix_wall_lj93_kokkos.cpp (hunk 4 above)
Comment (alphataubio) [reply src/KOKKOS/fix_wall_lj93_kokkos.cpp:40]: there is only one parallel region in FixWallLJ93Kokkos, then it is fine to set datamask_read. no change needed here.

File: src/KOKKOS/fix_wall_lj93_kokkos.cpp (hunk 5 above)
Comment (alphataubio) [reply src/KOKKOS/fix_wall_lj93_kokkos.cpp]: fixed by commit https://github.com/lammps/lammps/pull/4190/commits/cac6b5dc02afcff7b42c96a90cf829a4da537a2a

File: src/KOKKOS/fix_wall_lj93_kokkos.cpp (hunk 6 above)
Comment (alphataubio) [reply src/KOKKOS/fix_wall_lj93_kokkos.cpp]: fixed by commit https://github.com/lammps/lammps/commit/cac6b5dc02afcff7b42c96a90cf829a4da537a2a

File: src/fix_wall.cpp (hunk 7 above)
Comment (alphataubio) [reply src/fix_wall.cpp]: fixed by commit https://github.com/lammps/lammps/commit/cac6b5dc02afcff7b42c96a90cf829a4da537a2a

File: src/KOKKOS/fix_efield_kokkos.cpp (hunk 2 above)
Comment (alphataubio) [reply src/KOKKOS/fix_efield_kokkos.cpp:242]: it does now. this was my first attempt at debugging failed test but instead it was a symptom of another issue https://github.com/lammps/lammps/issues/4257.

along the way i ended up:
- fixing variable case (TagFixEfieldNonConstant operator) which didnt work anyways in kokkos class
- adding `fix_modify virial yes` support to match documentation, which was missing in kokkos class
- adding dipole support in kokkos class after i noticed your dipole code in base class, which was missing in kokkos class

**Comment by alphataubio** [comment]
Also i ran the unit tests with GoogleTest [Undefined Behavior Sanitizer](https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html), [Address Sanitizer](https://github.com/google/sanitizers/wiki/AddressSanitizer), and [Thread Sanitizer](https://github.com/google/sanitizers/wiki/ThreadSanitizerCppManual). **No issues detected**

see https://github.com/lammps/lammps/issues/4229#issuecomment-2268817624

**Comment by alphataubio** [comment]
@stanmoore1 i tested `fix efield/kk` on both 1 and 4 GPUs. It works, I was concerned about comment in the code about crash accessing unwrap[] but it doesn't and numerics are similar to serial

File: src/fix_efield.h [hunk 8]
```diff
@@ -59,7 +59,7 @@ class FixEfield : public Fix {
   double **efield;
 
   int force_flag;
-  double fsum[4], fsum_all[4];
+  double *fsum, fsum_all[4]; // need fsum double*, not double[] for kokkos dual view
```
Comment (stanmoore1) [inline src/fix_efield.h]: Is this change still needed?

**Review by stanmoore1 — COMMENTED** [review]
@alphataubio overall this looks good. I had one question in the comments.

**Comment by alphataubio** [comment]
> @alphataubio overall this looks good. I had one question in the comments.

No I don't need fsum in Kokkos subclass anymore 

Cleanup commits done, unit tests all pass

**Comment by stanmoore1** [comment]
@alphataubio thanks, this is good work. I'll run it through my regression testing, then it should be ready to merge.

**Comment by alphataubio** [comment]
It still needs a review by @sjplimp because I added functions to base classes 

but no rush next stable is not for another year. 

If @stanmoore1 you see people having problems with these you can just tell them to build develop plus this PR.

**Comment by stanmoore1** [comment]
Running Kokkos regression tests on this PR now...

**Comment by stanmoore1** [comment]
A couple regression tests are failing with CUDA:

```
11:32:36 dir = /lammps_testing/examples/for_kokkos/reax/water
11:32:36 test = water.qeq.field
11:32:36 gold standard = log.23Sep24.linux.kk_gpu_2.water.qeq.field
11:32:51 ==== comparing log.kk_gpu_2.water.qeq.field with log.23Sep24.linux.kk_gpu_2.water.qeq.field ====
11:32:51 mpiexec noticed that process rank 1 with PID 0 on node cee-h100-002 exited on signal 6 (Aborted).
11:32:51 !!! test water.qeq.field FAILED
11:32:51 elapsed time = 14.8613550663 s for test water.qeq.field
```

```
 *** Process received signal ***
Signal: Aborted (6)
Signal code:  (-6)
(ptr->cuda_stream_synchronize_wrapper(stream)) error( cudaErrorIllegalAddress): an illegal memory access was encountered ../../lib/kokkos/core/src/Cuda/Kokkos_Cuda_Instance.cpp:165
```

**Comment by alphataubio** [comment]
> A couple regression tests are failing with CUDA:
> 
> ```
> 11:32:36 dir = /lammps_testing/examples/for_kokkos/reax/water
> 11:32:36 test = water.qeq.field
> 11:32:36 gold standard = log.23Sep24.linux.kk_gpu_2.water.qeq.field
> 11:32:51 ==== comparing log.kk_gpu_2.water.qeq.field with log.23Sep24.linux.kk_gpu_2.water.qeq.field ====
> 11:32:51 mpiexec noticed that process rank 1 with PID 0 on node cee-h100-002 exited on signal 6 (Aborted).
> 11:32:51 !!! test water.qeq.field FAILED
> 11:32:51 elapsed time = 14.8613550663 s for test water.qeq.field
> ```
> 
> ```
>  *** Process received signal ***
> Signal: Aborted (6)
> Signal code:  (-6)
> (ptr->cuda_stream_synchronize_wrapper(stream)) error( cudaErrorIllegalAddress): an illegal memory access was encountered ../../lib/kokkos/core/src/Cuda/Kokkos_Cuda_Instance.cpp:165
> ```

ok ill have a look at it now

**Comment by stanmoore1** [comment]
@alphataubio the OpenMP test passes fine, so probably trying to access CPU memory the GPU, leading to a segfault.

**Review by stanmoore1 — APPROVED** [review]
Kokkos regression tests passed.

**Comment by stanmoore1** [comment]
@akohlmey this PR has some bugfixes that were required to get the unit tests to pass when using Kokkos, I think we need to backport all the changes in `/src` to the next stable update. Nothing outside of `/src` (e.g. `/unittest`) needs to be backported.

**Comment by stanmoore1** [comment]
@alphataubio one question:
```
pair_coul_dsf_kokkos INTENTIONALY SKIPPED IN YAML FILE
pair_zbl_kokkos INTENTIONALY SKIPPED IN YAML FILE
```

Why were these skipped?

**Comment by alphataubio** [comment]
> @alphataubio one question:
> 
> ```
> pair_coul_dsf_kokkos INTENTIONALY SKIPPED IN YAML FILE
> pair_zbl_kokkos INTENTIONALY SKIPPED IN YAML FILE
> ```
> 
> Why were these skipped?

it wasnt me. i documented that `skip_tests: kokkos_omp` from yaml files for those two pairs by @akohlmey 2 years ago: 

[skip tests that fail with Kokkos+OpenMP](https://github.com/lammps/lammps/commit/35a4c0f5010e877385b1fa0deb610681d4d66736)
+6 −1 7 lines changed
[Add force style tests for Kokkos... #3035](https://github.com/lammps/lammps/pull/3035)
[akohlmey](https://github.com/lammps/lammps/commits?author=akohlmey) committed on Feb 7, 2022

let me turn those back on to see what happens...

**Comment by akohlmey** [comment]
> @akohlmey this PR has some bugfixes that were required to get the unit tests to pass when using Kokkos, I think we need to backport all the changes in `/src` to the next stable update. Nothing outside of `/src` (e.g. `/unittest`) needs to be backported.

@stanmoore1 changes should now be in the maintenance branch (PR #4327). Already tested locally with Kokkos_OMP.

Please note, I had to restore the changes for my Radeon 780M chip and HIP back as they were overwritten when updating Kokkos. https://github.com/lammps/lammps/pull/4190/commits/db4c6fbaaae72033815811f32777262edc60c91f

**Comment by stanmoore1** [comment]
Thanks @akohlmey!

**Comment by stanmoore1** [comment]
Windows unit test failure seems unrelated to the changes in this PR:
```
Failures
 - ninja (exited 1) - ninja not installed. An error occurred during installation:
 Error downloading 'ninja.1.12.1' from 'https://community.chocolatey.org/api/v2/package/ninja/1.12.1'.
Error: Process completed with exit code 1.
```

**Comment by akohlmey** [comment]
> Windows unit test failure seems unrelated to the changes in this PR:
> 
> ```
> Failures
>  - ninja (exited 1) - ninja not installed. An error occurred during installation:
>  Error downloading 'ninja.1.12.1' from 'https://community.chocolatey.org/api/v2/package/ninja/1.12.1'.
> Error: Process completed with exit code 1.
> ```

Yes, this is an unfortunate effect I am seeing now that we are using GitHub runners more. Sometimes they are disconnected from external servers, e.g. to download ubuntu packages, and then we get failures galore. Much less than with the Temple machines.

**Comment by akohlmey** [comment]
> Windows unit test failure seems unrelated to the changes in this PR

BTW: You should be able to re-run the test after clicking on "Details".

**Comment by akohlmey** [comment]
For the record. The following force-style tests have the flag to skip the tests using KokkosOpenMP:
```
force-styles/tests/improper-class2.yaml
force-styles/tests/kspace-ewald_conp_charge.yaml
force-styles/tests/kspace-pppm_ad.yaml
force-styles/tests/kspace-pppm_conp_charge.yaml
force-styles/tests/manybody-pair-pod.yaml
force-styles/tests/manybody-pair-tersoff_mod_shift.yaml
force-styles/tests/manybody-pair-tersoff_shift.yaml
force-styles/tests/manybody-pair-tersoff_zbl_shift.yaml
force-styles/tests/mol-pair-coul_dsf.yaml
force-styles/tests/mol-pair-dpd_coul_slater_long.yaml
force-styles/tests/mol-pair-dpd_ext_tstat.yaml
force-styles/tests/mol-pair-dpd_ext.yaml
force-styles/tests/mol-pair-dpd_tstat.yaml
force-styles/tests/mol-pair-dpd.yaml
force-styles/tests/mol-pair-hybrid_multiple.yaml
force-styles/tests/mol-pair-hybrid-overlay.yaml
force-styles/tests/mol-pair-hybrid-scaled.yaml
force-styles/tests/mol-pair-lj_cut_coul_dsf.yaml
force-styles/tests/mol-pair-table.yaml
force-styles/tests/mol-pair-zbl.yaml
```
Some are principal problems, others require detailed debugging (and we are not even testing GPUs or OpenMP with different settings for neighbor list etc. so there may be more issues looming).

After all the changes in this PR, I think it would be better to merge it and focus on fixing tests later. The bigger challenge at the moment is to get the new regression tests working well.
